cajita/src/Cajita_FastFourierTransform.hpp

```diff
@@ -28,232 +28,180 @@ namespace Cajita
-        _fft.scaled = params.scaled;
-
-        // Get the global grid.
-        const auto &global_grid = layout.localGrid()->globalGrid();
+        // Set FFT options from given parameters
+        _fft = std::make_shared<heffte::fft3d<backend_type>>(
+            inbox, outbox, layout.localGrid()->globalGrid().comm(),
+            heffte_params );
 
-        // Get the global dimensions of the problem. K indices move the
-        // fastest because we fix the work array to be layout right.
-        std::array<int, 3> global_num_entity = {
-            global_grid.globalNumEntity( EntityType(), Dim::K ),
-            global_grid.globalNumEntity( EntityType(), Dim::J ),
-            global_grid.globalNumEntity( EntityType(), Dim::I )};
+        int fftsize = std::max( _fft->size_outbox(), _fft->size_inbox() );
 
-        // Get the local dimensions of the problem.
+        // Check the size.
         auto entity_space =
             layout.localGrid()->indexSpace( Own(), EntityType(), Local() );
-        std::array<int, 3> local_num_entity = {
-            (int)entity_space.extent( Dim::K ),
-            (int)entity_space.extent( Dim::J ),
-            (int)entity_space.extent( Dim::I )};
-
-        // Get the low corner of the global index space on this rank.
-        std::array<int, 3> global_low = {
-            (int)global_grid.globalOffset( Dim::K ),
-            (int)global_grid.globalOffset( Dim::J ),
-            (int)global_grid.globalOffset( Dim::I )};
-
-        // Get the high corner of the global index space on this rank.
-        std::array<int, 3> global_high = {
-            global_low[Dim::I] + local_num_entity[Dim::I] - 1,
-            global_low[Dim::J] + local_num_entity[Dim::J] - 1,
-            global_low[Dim::K] + local_num_entity[Dim::K] - 1};
-
-        // Setup the fft.
-        int permute = 0;
-        int fftsize, sendsize, recvsize;
-        _fft.setup( global_num_entity.data(), global_low.data(),
-                    global_high.data(), global_low.data(), global_high.data(),
-                    permute, fftsize, sendsize, recvsize );
-
-        // Check the size.
         if ( fftsize < (int)entity_space.size() )
-            throw std::logic_error( "HEFFTE expected allocation size smaller "
+            throw std::logic_error( "Expected FFT allocation size smaller "
                                     "than local grid size" );
 
-        // Allocate the work array.
-        _fft_work = Kokkos::View<Scalar *, DeviceType>(
-            Kokkos::ViewAllocateWithoutInitializing( "fft_work" ),
-            2 * fftsize );
+        _fft_work = Kokkos::View<std::complex<Scalar> *, DeviceType>(
+            Kokkos::ViewAllocateWithoutInitializing( "fft_work" ), fftsize );
     }
 
     /*!
       \brief Do a forward FFT.
       \param in The array on which to perform the forward transform.
     */
     template <class Array_t>
-    void forward( const Array_t &x )
+    void forward( const Array_t &x, const FFTScaleNone )
```

Comment:
These should be an `override` implementation of a pure virtual function in the base class. That way future FFT implementations have to conform to the same interface.

Reply:
That was my plan, but how would I do that with the template?

Reply:
explicit overloads with each template variant

Reply:
At your suggestion I used CRTP for this. It's a much better design, but I couldn't see how to simplify the template parameters of the derived class like you mentioned.